app/drivers/kscan/debounce.c

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2021 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "debounce.h"
+
+static uint32_t get_threshold(const struct debounce_state *state,
+                              const struct debounce_config *config) {
+    return state->pressed ? config->debounce_release_ms : config->debounce_press_ms;
+}
+
+void debounce_update(struct debounce_state *state, const bool active,
+                     const struct debounce_config *config) {
+    // This uses a variation of the integrator debouncing described at
+    // https://www.kennethkuhn.com/electronics/debounce.c
+    // Every update where "active" does not match the current state, we increment
+    // a counter, otherwise we decrement it. When the counter reaches a
+    // threshold, the state flips and we reset the counter.
+    if (active == state->pressed) {
+        if (state->counter > 0) {
+            state->counter--;
+        }
+        return;
+    }
+
+    const uint32_t flip_threshold = get_threshold(state, config);
+
+    if (state->counter < flip_threshold) {
+        state->counter++;
+        return;
+    }
+
+    state->pressed = !state->pressed;
+    state->counter = 0;
+    state->changed = true;
```

Comment:
I would suggest not to return in the middle of the function, but to employ else statements instead.

```suggestion
    if (active == state->pressed) {
        if (state->counter > 0) {
            state->counter--;
        }
    } else {
        const uint32_t flip_threshold = get_threshold(state, config);
  
        if (state->counter < flip_threshold) {
            state->counter++;
        } else {
            state->pressed = !state->pressed;
            state->counter = 0;
            state->changed = true;
        }
    }
```

Reply:
> As an example, consider a function that uses dynamic memory allocation and you free at the end, a premature return skips the freeing step and breaks the functionality. You would have to replicate it before each of the exit points of your function.

That is a very good argument (and one I agree with) for only having a single return point in a function that allocates resources, but this function doesn't (and likely never will).

This function is simple enough that I think it's pretty readable either way, but I personally try to avoid having many levels of nested control structures. With a return, I know the function is done and don't have to check if there's any code after the if/else block.

Reply:
> but this function doesn't (and likely never will)

That is true. The function is indeed simple enough to be readable this way.

I agree that the early return makes handling of if/else blocks simpler and possibly more robust.

Reply:
Good nuance here, would agree with the opinion about *this* function that the early returns lead to clearer code.